File: Hyerin Cheon/2022-07-02-Word-Break.js
var wordBreak = function (s, wordDict) {
  let dp = new Array(s.length + 1).fill(false);
  dp[0] = true;

  for(let i = 0; i <= s.length; i++){
    if(dp[i] === true){
      for(let word of wordDict){
        if(s.slice(i, i + word.length) === word){
          dp[i + word.length] = true;
        }
      }
    }
  }
  return dp[s.length];
};

File: Hyonsok Na/week4/2022-07-02-Word-Break.java
class Solution 
{
  public boolean wordBreak(String s, List<String> wordDict )
  {
      // O(n^3) time | O(n) space
      if(s == null || s.length() == 0) return false;

      int n = s.length();

      // dp[i] represents whether s[0...i] can be formed by dict
      boolean[] dp = new boolean[n];
      HashSet<String> mySet = new HashSet<>(wordDict);

      for (int i = 0; i < n; i++) 
      {
          for (int j = 0; j <= i; j++) 
          {
              String sub = s.substring(j, i + 1);
    
              if(mySet.contains(sub) && (j == 0 || dp[j - 1])) 
              {
                  dp[i] = true;
                  break;
              }
          }
      }
      return dp[n - 1];
  }
}